Fix mod browser previews

RoosterDragon opened the pull request:
Fixes regression from #7919, which caused invalid previews on some of the mod browser images.

reaperrr commented:
Can confirm that this fixes the previews for me, but this still happens (both on bleed and this branch) when I wanted to choose a different map in TD mod (skirmish):

Tiberian Dawn Mod at Version {DEV_VERSION}
Operating System: Windows (Microsoft Windows NT 6.1.7601 Service Pack 1)
Runtime Version: .NET CLR 4.0.30319.18444
Exception of type `System.Collections.Generic.KeyNotFoundException`: Der angegebene Key war nicht im Dictionary angegeben.
   bei System.Collections.Generic.Dictionary`2.get_Item(TKey key)
   bei OpenRA.ReadOnlyDictionary`2.get_Item(TKey key) in e:\OpenRA_work2\OpenRA.Game\Primitives\ReadOnlyDictionary.cs:Zeile 77.
   bei OpenRA.MapCache.LoadAsyncInternal() in e:\OpenRA_work2\OpenRA.Game\Map\MapCache.cs:Zeile 176.
   bei System.Threading.ThreadHelper.ThreadStart_Context(Object state)
   bei System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   bei System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   bei System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state)
   bei System.Threading.ThreadHelper.ThreadStart()

penev92 commented on OpenRA.Mods.Common/Widgets/Logic/ModBrowserLogic.cs:
```diff
@@ -96,6 +84,23 @@ public ModBrowserLogic(Widget widget)
+			{
+				using (var image = new Bitmap(path))
+					if (image.Size == requiredSize)
+					{
```
The `if` has `{ }`, so OpenRA regulations state that the outer `using` must have as well.

RoosterDragon replied:
Amended.

penev92 commented:
Code looks good, and it works. :+1: 

I'm not going to merge yet because I have a question: isn't this just hiding a problem with the images that would be better resolved by fixing the images themselves than hiding any problems? (the 2 images were obviously in a format that OpenRA doesn't like)
Also the image conversion for incompatible images will be ran every time the mod chooser loads. The game can instead fix the preview images itself (save the fixed image over the original).

reaperrr commented:
The problem appears to be that OpenRA wants PNGs in 32bpp (now at least, before #7919 24bpp worked fine), while some editing programs (applies at least to IrfanView and MS Paint) save PNGs in 24bpp.

RoosterDragon commented:
Yeah we needed #7919 because a particular 24bbp custom map preview failed on mono. It seems to work most of the rest of the time though.

So we have a few options:
1) Have ```sheetBuilder.Add``` perform a conversion internally if the input is not right.
2) Have callers of ```sheetBuilder.Add``` only be allowed to pass stuff in the right format.
3) Have all game images be forced to be 32bbp.

Options 3 seems dumb to me. I don't want to force people to save their image in a certain bit depth just because we're too lame to handle that.

Option 1 and 2 are both fine by me.

penev92 commented:
Converting all images to the correct format at runtime each time the game loads seems pretty crap to me. (which both 1) and 2) do in case there are images with incorrect format)
I have to say I only like 3)
Being consistent is a good thing, and it shouldn't be a problem for modders to save their mods' images in the proper format. Also, in case they can't for some reason, the game can convert them itself, so how about having it convert and overwrite images that are in a different format? That way the next time it runs all will be in order.

pchote commented:
My preference here is 1 (for mod compatibility) plus converting any assets that aren't 32bpp in our default mods (to avoid unnecessary work).

RoosterDragon committed:
Convert to 32bbp internally in FastCopyIntoSprite.

This avoids the need for callers to ensure the pixel format is correct, but ensures that the copying succeeds when the format is different.

RoosterDragon commented:
OK #7961 has ensured our provided images are in 32bbp. I have rewritten this PR to follow option 1, and convert to 32bbp just-in-time within the ```FastCopyIntoSprite``` method which will handle things for mod or map previews not in our control.

penev92 commented:
Looks OK to me :+1:

obrakmann commented:
I reverted 8b07c08 for testing and can confirm that this works as intended. Code looks good, too. :+1:

obrakmann commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog/_compare/9eefbddf04503290d1e9d5c71b7c5a42dbe5b79e%5E...9eefbddf04503290d1e9d5c71b7c5a42dbe5b79e)

obrakmann commented:
err... this worked for the mod previews, but it does not seem to work for the map previews in ra, which are all RGB images. (the cnc ones are all RGBA already). It seems there's a check left in `MapCache.cs` which throws away RGB images. I'll whip up a patch.
